lib/message_queue.cpp

```diff
@@ -99,28 +99,28 @@ MessageHandler::MessageHandler() : running(true) {
                     if (data.at("type") == "result") {
                         // only deliver result to handler with matching id
                         if (handler_->id == data.at("data").at("id")) {
-                            handler(data.at("data"));
+                            handler(message.topic, data.at("data"));
                         }
                     }
                 } else if (handler_->type == HandlerType::SubscribeVar) {
                     // unpack var
                     if (handler_->name != data.at("name")) {
                         continue;
                     }
-                    handler(data.at("data"));
+                    handler(message.topic, data.at("data"));
                 } else {
                     // external or unknown, no preprocessing
-                    handler(data);
+                    handler(message.topic, data);
                 }
             }
         }
     });
 }
 
-void MessageHandler::add(std::shared_ptr<json> message) {
+void MessageHandler::add(MessageDetails message) {
     {
         std::lock_guard<std::mutex> lock(this->handler_ctrl_mutex);
-        this->message_queue.push(message);
+        this->message_queue.emplace(message);
```

Comment:
Use `std::move(message)` here, otherwise the topic string will get copied.

Reply:
updated to use push and move.